[FIXED JENKINS-10629] - Support of big files archiving (>8Gb)

oleg-nenashev opened the pull request:
[JENKINS-10629](https://issues.jenkins-ci.org/browse/JENKINS-10629)

Summary:
- Migrate the core to commons-compress instead of the local implementation
  - According to the investigation, local changes in <code>TarInputStream</code> have been fixed in the original source as well
- Enable BigNumber mode (STAR implementation, which uses GNU extensions)

I've tested the change locally, but I'm not sure if it makes sense to introduce such kind of tests in unit or harness test suites (high memory utilisation, tmp files for harness).

@reviewbybees

oleg-nenashev commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -2266,12 +2267,15 @@ private Integer writeToTar(File baseDir, DirScanner scanner, OutputStream out) t
 
     /**
      * Reads from a tar stream and stores obtained files to the base dir.
+     * @since TODO: the method uses 
```
Incomplete text. For noting: "The method uses classes from commons-compress instead of the home-made implementation"

jglick replied:
You do not need `@since` on non-API elements.

oleg-nenashev replied:
In such case <code>@since</code> allows to track major changes, which may be important to core developers. BTW, no problem if we remove this annotation

KostyaSha replied:
+1 for keeping `@since` Since is valid javadoc and defines from what version appeared

jglick replied:
Harmless though I would normally expect to use Git history to determine exactly when an implementation was changed.

KostyaSha replied:
Any needed information can be mentioned in javadoc description, you can ignore it and continue wasting your time in git history.
since - when first appeared. In description <p> since 1.XXX.2 - supports large file > 10 GB

KostyaSha commented on core/pom.xml:
```diff
@@ -277,6 +277,11 @@ THE SOFTWARE.
     </dependency>
     <dependency>
+        <groupId>org.apache.commons</groupId>
+        <artifactId>commons-compress</artifactId>
```
Wouldn't it start conflicting with some existed plugin?

oleg-nenashev replied:
The only dependency declarer in jenkinsci or is https://github.com/jenkinsci/p4-plugin. It declares the same version requirement, hence @p4paul can easily remove the dependency when he upgrades to the new core version.

KostyaSha commented:
+1 for using not self implemented algo

jglick commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -2280,8 +2284,7 @@ private void readFromTar(String name, File baseDir, InputStream in) throws IOExc
 
-                    byte linkFlag = (Byte) LINKFLAG_FIELD.get(te);
-                    if (linkFlag==TarEntry.LF_SYMLINK) {
+                    if (te.isSymbolicLink()) {
```
Can delete `LINKFLAG_FIELD` and `getTarEntryLinkFlagField`.

oleg-nenashev replied:
https://github.com/oleg-nenashev/jenkins/commit/3106f9c25be35acb05086548ede32118376bcd33  deletes both entries

jglick replied:
No, that deletes `LINKNAME_FIELD`. https://github.com/oleg-nenashev/jenkins/blob/JENKINS-10629/core/src/main/java/hudson/FilePath.java#L2726-2738

oleg-nenashev replied:
@jglick 
Probably something is wrong with my Git clients, but I definitely see that <code> getTarEntryLinkFlagField()</code> has been deleted as well

jglick replied:
Well it does not show up in the GH diff. Maybe you forgot to push something?

oleg-nenashev replied:
Most probably we're talking about different code sections:
![screen shot 2015-04-14 at 17 14 32](https://cloud.githubusercontent.com/assets/3000480/7138578/edc8762c-e2c9-11e4-9940-64a185b2d569.png)

oleg-nenashev replied:
@jglick [:facepalm:], please ignore

oleg-nenashev commented:
Seems that PR builder is offline.

jglick commented:
> Seems that PR builder is offline.

Yes, sorry, I was going to look into this.

:+1: other than the nits noted above.

KostyaSha commented:
@oleg-nenashev known issue #INFRASUCKS https://groups.google.com/forum/#!topic/jenkinsci-dev/ZDtREZpH8dQ

jglick commented:
> I'm not sure if it makes sense to introduce such kind of tests in unit or harness test suites (high memory utilisation, tmp files for harness)

Agreed that we would not want that kind of test run normally. I would suggest writing the functional test—just a project with a builder that creates a 10Gb file and archives it—but marking it `@Ignore`. If someone wants to investigate problems later on, they need merely comment out the annotation. (Or JUnit may have some annotation understood by the default test runner that lets the test be activated via system property; cannot remember offhand.)

KostyaSha commented:
You can try create test case with `fallocate` that should be fast for creation and archiving.

KostyaSha commented on core/src/main/java/hudson/org/apache/tools/tar/TarInputStream.java:
```diff
@@ -37,10 +37,10 @@
- *
+ * @deprecated Use {@link org.apache.commons.compress.archivers.tar.TarArchiveInputStream} instead
  */
 public class TarInputStream extends FilterInputStream {
```
And `@Deprecated`

KostyaSha commented on core/src/main/java/hudson/org/apache/tools/tar/TarOutputStream.java:
```diff
@@ -35,6 +35,8 @@
+ * @deprecated Use {@link org.apache.commons.compress.archivers.tar.TarArchiveOutputStream} instead
  *
  */
 public class TarOutputStream extends FilterOutputStream {
```
`@Deprecated`

oleg-nenashev commented:
I suppose the PR is ready for the merge. BTW, a final confirmation from PRB would be useful

jtnord commented on core/src/test/java/hudson/FilePathTest.java:
```diff
@@ -325,6 +328,33 @@ private FilePath createFilePath(final File base, final String... path) throws IO
+    @Ignore
+    @Test public void archiveBigFile() throws Exception {
+        assumeTrue(!Functions.isWindows());
+        long largeFileSize = 10000000000l;
```
is this 10000000000l or 100000000001?
Use upper case 'L'
a lower case 'L' should not be used for Longs or variable names.

jtnord commented on core/src/test/java/hudson/FilePathTest.java:
```diff
@@ -325,6 +328,33 @@ private FilePath createFilePath(final File base, final String... path) throws IO
+        tempFile.deleteOnExit();
+ 
+        file.setLength(largeFileSize);
+        assumeTrue(largeFileSize == file.length());
```
file is not closed.

jtnord commented on core/src/test/java/hudson/FilePathTest.java:
```diff
@@ -325,6 +328,33 @@ private FilePath createFilePath(final File base, final String... path) throws IO
+ 
+        int tar = new FilePath(tempFile.getParentFile()).tar(new FileOutputStream(tarFile), tempFile.getName());
+ 
+        assertEquals("One tar file should have been compressed", 1, tar);
```
you should not wait for JVM termination to tidy up after tests.

jtnord replied:
nit: its not the tar file that is being compressed (and nothing is being compressed).  Its one file file being archived (or tarred)

jtnord commented on core/src/test/java/hudson/FilePathTest.java:
```diff
@@ -325,6 +328,33 @@ private FilePath createFilePath(final File base, final String... path) throws IO
+    @Issue("JENKINS-10629")
+    @Ignore
+    @Test public void archiveBigFile() throws Exception {
+        assumeTrue(!Functions.isWindows());
```
Why - this will run on Windows as well as well.  It is only run when a user explicitly wants to run it.
It's not like the Linux machine is checked for sparse file support first.

jtnord replied:
For giggles I just timed creation of a file of size  10000000000 bytes on 'Doze 8.1 on NTFS and it took 0.52ms.

jtnord commented:
:-1:  happy with the production code - but the test has several issues.

oleg-nenashev commented:
@jtnord 
Thanks for the review. I agree with the most of comments regarding the test case. I'll rework it

oleg-nenashev commented:
@jtnord, I've completely reworked the test. Now it implements a roundtrip check, which is being used for zero/small files (enabled) and big files (disabled by default).

jtnord commented:
:+1:

jglick commented on core/src/test/java/hudson/FilePathTest.java:
```diff
@@ -325,6 +328,67 @@ private FilePath createFilePath(final File base, final String... path) throws IO
+        FilePath outDir = null;
+        
+        try {
+            tempFile = File.createTempFile(filePrefix, ".log");
```
You can just use `FilePathTest.temp` to allocate temporary files without bothering with a `finally`-block.

oleg-nenashev committed:
[JENKINS-10629] - Direct unit test for the issue

This test also introduces additional round-trip tests for small files

oleg-nenashev committed:
[JENKINS-10629] - Migrate the Tar archives handling code to commons-compress

oleg-nenashev committed:
FIXED JENKINS-10629] - Enable BigNumber mode to support archiving of files with size >8Gb

BIGNUMBER_STAR mode has been selected, because it utilizes GNU extensions (like LONGFILE_GNU does).

oleg-nenashev commented:
The pull request build hanged on <code>lib.layout.RenderOnDemandTest</code>.

jglick commented:
:+1: 

Anyone merging the PR should do so via validated merge to ensure that the tests get run.

KostyaSha commented on core/src/main/java/hudson/util/io/TarArchiver.java:
```diff
@@ -58,13 +60,14 @@
-        });
-        tar.setLongFileMode(TarOutputStream.LONGFILE_GNU);
+        });    
+        tar.setBigNumberMode(TarArchiveOutputStream.BIGNUMBER_STAR);
```
Why not  BIGNUMBER_POSIX ?

oleg-nenashev replied:
See the commit message. BIGNUMBER_STAR uses GNU tar extensions

oleg-nenashev commented:
https://github.com/jenkinsci/jenkins/pull/1667 temporarily reverts changes till we have a fix for the binary stream corruption issue (and dozens of unit tests)
